### PR TITLE
T4rec use case

### PR DESCRIPTION
**Pull request by radekosmulski**
This is a PR that implements https://github.com/NVIDIA-Merlin/Merlin/issues/694 (what is currently possible)

### CODE REVIEW

File: examples/usecases/transformers-next-item-prediction.ipynb [hunk 1]
```diff
@@ -0,0 +1,910 @@
+{
```
Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.    get_booking('./data')
</pre><p> </p><p>do you mind to give the absolute path like <code>/workspace/data</code> instead of <code>.data</code> ?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    itineraries = get_lib().read_csv('./data/train_set.csv', parse_dates=['checkin'])
</pre><p>shall we keep the dataframe name as <code>train</code> ? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>I believe since this is still based on HF architectures, we need to provide a max_length, and under the hood we still pad the sequences, even duh we do not pad it in the preprocessing. Can you confirm with Sara?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>can you add a sentence above to explain what you do in this cell brielfy? something like "let's check number of unique trips` . </p><p> </p><p>also do you need to do <code>sample(frac=1)</code> for that? can you not do <code>itineraries.utrip_id.nunique()</code> directly? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    
</pre><p>can you add a sentence and explain what type of split you are doing here? like <code>first 80% of the data is set as train, and the rest is set as validation</code>  and data is sorted by checkin time? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>can  you rewrite nvtabular as NVTabular? May be rephrase <code>If you would like to learn more about NVTabular  library, ...</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    from merlin.models.tf import Loader
</pre><p>Guess some imports repeats again? you can put all the imports in the first cell may be?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    validation_set_dataset = Dataset(validation_set)
</pre><p>where is Dataset  imported? I guess import code is missing?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    validation_set_processed = wf.fit_transform(validation_set_dataset)
</pre><p>we dont use <code>fit</code> on validation set. it should be only <code>transform</code>.   so just this: <code>wf.transform(validation_set_dataset)</code>.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.    )
</pre><p>can we add more sequential and non-sequential features?</p><p> </p><p>like</p><p> </p><p><code>"device_class": ["first"],</code></p><p>"hotel_country": ["list"],</p><p>&nbsp;&nbsp;"weekday_checkin": ["list"],</p><p>&nbsp;"weekday_checkout": ["list"],</p><p> </p><p>you can generate temporal features like below but note that <code>checkin  should be datetime object first.</code></p><p> </p><pre class="ql-syntax" spellcheck="false">
weekday_checkin = (
&nbsp; &nbsp; ["checkin"]
&nbsp; &nbsp; &gt;&gt; nvt.ops.LambdaOp(lambda col: col.dt.weekday)
&nbsp; &nbsp; &gt;&gt; nvt.ops.Rename(name="weekday_checkin")
)


weekday_checkout = (
&nbsp; &nbsp; ["checkout"]
&nbsp; &nbsp; &gt;&gt; nvt.ops.LambdaOp(lambda col: col.dt.weekday)
&nbsp; &nbsp; &gt;&gt; nvt.ops.Rename(name="weekday_checkout")
)


</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (rnyak) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>can you explain here some args with one sentence?  what's this <code>pre=mm.ReplaceMaskedEmbeddings()</code> doing for example? Can you pls explain that under the hood this is doing masked language modeling and masking the items randomly? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>Interestingly, going <code>/workspace/data</code> throws an error. I think there might be an issue in the way getting the data is implemented. Let me check with Marc </p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>change made</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>reached out to Sara to confirm</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>Sentence added</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>absolutely, explanation added</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>good ploint, removed duplicates and moved it all to first cell</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>this was imported as part of <code>from nvtabular import *</code> I believe </p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>fixed</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>added additional features</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>explanation added</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>checked with Sara, based on the response I believe this is okay, here is the description of how all these pieces fit together in MMs:</p><p> </p><blockquote>"In Merlin models, we removed the max-length parameter and relies on the tf.ragged properties to automatically pad the sequences of each batch before applying the HF transformer layer.</blockquote><p>In fact, converting a ragged to a dense tensor will automatically pad the sequences of a given batch to the maximum length found in this batch. So the max length is also variable from one batch to another"</p>

**Comment by radekosmulski** [comment]
rerun tests

File: examples/usecases/transformers-next-item-prediction.ipynb (hunk 1 above)
Comment (gabrielspmoreira) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>I think you need to add the <code>sort_cols="checkin" 
</code>arg to the <code>nvt.ops.Groupby</code> op , to ensure that each trip is sorted by check-in, as done in the Groupby op in this <a href="https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/examples/end-to-end-session-based/01-ETL-with-NVTabular.ipynb" target="_blank">example</a></p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: Thank you very much for this @gabrielspmoreira! I've made the change

File: examples/usecases/transformers-next-item-prediction.ipynb [hunk 2]
```diff
@@ -0,0 +1,1013 @@
...
```
[the remark this answers is not in the document]
Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: Thank you very much @gabrielspmoreira! 🙏 Added this information to the notebook, much appreciated!

File: examples/usecases/transformers-next-item-prediction.ipynb [hunk 3]
```diff
@@ -0,0 +1,1234 @@
...
```
Comment (bschifferer) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>We need to remove this</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (bschifferer) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>We can add a new headline with Preprocessing with NVTabular?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (bschifferer) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>Can we have all imports in the beginning of the notebook?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (bschifferer) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>Do we need the Loader? In previous Merlin Models, we could feed the nvt.Dataset object to model.fit and it will construct the loader automatically?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (bschifferer) [inline examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>can we change this to return dictonary that we know what these numbers are?</p><p> </p><p>Do we have any Conclusion or Next Steps?</p><p> </p><p>E.g. Next steps could be dressipi example?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/853/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

**Comment by bschifferer** [comment]
The example is great! I added some minor comments.

File: examples/usecases/transformers-next-item-prediction.ipynb (hunk 3 above)
Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>done</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>headline added added</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>absolutely, moved this to the beginning</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>you are absolutely right! turns out we don't need to be creating the loader, everything works perfectly fine if we just pass the dataset, removed this (and the creation of a loader for evaluation)</p>

Comment (radekosmulski) [reply examples/usecases/transformers-next-item-prediction.ipynb:1]: <p>Absolutely, that is a great idea to link to that notebook! I added another section where I am referencing the dressipi example.</p>

**Review by bschifferer — APPROVED** [review]
